### PR TITLE
Put

### LINKED ISSUE #64: `set` modifies the original object
The [readme for `set`](https://github.com/tjmehta/101#set) says:

> returns a new obj with the key and value set

I'd seen it as a powerful functionality which allowed me to get immutability in JS without using [an obese library][i].
[i]: https://www.npmjs.com/package/immutable "Immutable"

– until I noticed that [it's the original object that gets changed and returned](https://github.com/tjmehta/101/blob/master/set.js#L37) after all.

Is it supposed to be so?

### PR DESCRIPTION
**Pull request by tomekwi**
Fixes #64

### CODE REVIEW

**Comment by tomekwi** [comment]
This is mostly the same as `set` – only it `clone`s the object before assigning.

I've decided against `put(obj)` and `put(obj, setObj)`, because I don't really see a use case for them at the moment.

And if we now lock these options out, we have them open for the future when we find a real use case.

**Comment by stoeffel** [comment]
I hope I will have time on the weekend to review this (I'm in the military
this month and don't have a lot of computer time :-( ).

Tomek Wiszniewski <notifications@github.com> schrieb am Fr., 6. März 2015
11:55:

**Comment by tomekwi** [comment]
No hurry @stoeffel :)

Can't believe how dumb of me – I was sure Switzerland has no army except that in Vatican :)

**Comment by stoeffel** [comment]
Haha :smile:... No, unfortunately it's mandatory for everyone.

Tomek Wiszniewski <notifications@github.com> schrieb am Fr., 6. März 2015
13:21:

**Comment by tjmehta** [comment]
Awesome! So there are a few small questions that remain. Set supports objects in addition to key and val. Should put support objects as well?

put(obj1, obj2) and put(obj2)(obj1) ?

**Comment by tomekwi** [comment]
Thanks :)

As I said, I can throw these in as well if you think they're necessary. I just didn't see a use case for those, because as I see it:

```js
put(obj1, obj2)  ~=  assign(clone(obj1), obj2)
set(obj1, obj2)  ~=  assign(obj1, obj2)

put(obj)  ~=  assign.bind(null, {}, obj)
set(obj)  ~=  assign.bind(null, obj)
```

But I might be wrong :)

**Comment by stoeffel** [comment]
> I just didn't see a use case for those

I think so too, but I also think people would expect `put` to behave like `set`. So I think we should mirror those.

**Comment by tomekwi** [comment]
> people would expect put to behave like set

I totally agree. Just these are things which haven't been documented – I wasn't aware of those before. And I always see non-documented behavior as unreliable.

But to have both options open I've just pushed what you both wished :) @tjmehta you're free to merge whatever you think is right.

**Comment by stoeffel** [comment]
:+1:

**Comment by stoeffel** [comment]
How about using `set` in the `if`s to reduce duplicate code? Overkill?

**Comment by tomekwi** [comment]
Hmm, not a bad idea :)

But now I've looked at it:
- the [partial functions](https://github.com/tomekwi/101/blob/3ade/put.js#L26-L44) can't be handled by put
- the [error message](https://github.com/tomekwi/101/blob/3ade/put.js#L50-L52) is different

so there's not much left to save. Or do you have an idea for the partials as well?

**Comment by stoeffel** [comment]
> so there's not much left to save

I guess you are right... would propably be overkill.

**Comment by tjmehta** [comment]
i like `put` being it's own util. `set` and `del` are different type's of utils where we may want to support partial functionality *only* (bc `set(obj, key, val)` doesn't make any sense to use vs obj[key]=val). Put on the other hand could technically still be useful as `put(obj, key, val)`.

**Comment by stoeffel** [comment]
> put on the other hand could technically still be useful as put(obj, key, val).

Good point.

**Comment by stoeffel** [comment]
:bus:

**Comment by tomekwi** [comment]
:wave: @tjmehta is there anything I can improve yet?

File: put.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  if (arguments.length === 1) {
+    // (putObj)
+    putObj = obj;
+    if (!isObject(putObj)) throw new TypeError('Invalid arguments: expected putObj');
+    return function (obj) {
+      return assign(clone(obj), putObj); // returns new object
+    };
+  }
+  if (arguments.length === 2) {
+    if (isString(obj) || typeof obj === 'number') {
```
Comment (tjmehta) [inline put.js]: use `101/is-number`

**Comment by tjmehta** [comment]
1 comment, everything else looks good :+1: !

**Comment by tomekwi** [comment]
Thanks for reviewing – fixed. I picked that up [from `set`](https://github.com/tjmehta/101/blob/master/set.js#L30) – I've thought it's so for performance reasons.

**Comment by tjmehta** [comment]
Cool will change it there in another PR

**Comment by tjmehta** [comment]
101@0.14.0 - Thank you @tomekwi !

**Comment by tomekwi** [comment]
:beers: glad to help :)

**Pull request merged**